ocaml/typing/typeopt.mli

```diff
@@ -30,17 +30,30 @@ val bigarray_type_kind_and_layout :
       Env.t -> Types.type_expr -> Lambda.bigarray_kind * Lambda.bigarray_layout
 
 (* CR layouts: `layout` should have a `sort` argument.  *)
-(* CR layouts v2: [layout], [function_return_layout] and
-   [function2_return_layout] have had location arguments added just to support
-   the void check error message.  These arguments can be removed when we're
-   happy to take that check out. *)
-val layout : Env.t -> Location.t -> Types.type_expr -> Lambda.layout
+(* CR layouts v2: [layout], [function_return_layout], [function2_return_layout],
+   and [layout_of_sort] have had location arguments added just to support the
+   void check error message.  These arguments can be removed when we're happy to
+   take that check out. *)
+val layout :
+  Env.t -> Location.t -> Layouts.sort -> Types.type_expr -> Lambda.layout
+
+(* This translates a type system sort to a lambda layout.  Use the function
+   [layout] instead when the type is known: that function gives a more precise
+   result.  This should only be used when the precise layout isn't needed for
+   optimization. *)
+val layout_of_sort : Location.t -> Layouts.sort -> Lambda.layout
```

Comment:
This is unfortunate. I agree that it works, but I think we should try to remove this function. Leave a CR to return to this?

Reply:
I think the opposite: Computing a full `value_kind` is expensive and we often do it in places where the information can't be needed for optimization.  The middle-end team is working on calling `value_kind` less, not more, for that reason.  So we should be looking for more places where we can use this without impacting the performance of the generated code, not less.  I suspect there are a lot of them.

Reply:
I'm not suggesting to call `value_kind`, but instead to propagate the layout information. I agree that we should call `value_kind` less.

Put another way: How can we be sure that the full layout information isn't used for optimization? That seems like a very subtle property.

Reply:
I don't understand the first remark - the only reason to use `layout` instead of `layout_of_sort` is to get a call to `value_kind` - if we're not in the value case, the information they return is identical.

I agree the question of whether the detailed `value_kind` information is used is quite subtle, and is really for the middle end team to determine.  But they are actively engaged in that work, because calls to value_kind are a huge part of our compile time.

Reply:
I see one use of `layout_of_sort`, to prepare for a call to `lfunction`. Suppose a type is immediate. `layout_of_sort` will return `Pgenval`, not `Pintval`. That layout is then passed in `~params` to `lfunction`. Does the difference matter? It's hard to tell. The comment on `layout_of_sort` says that it should be used only when the precise layout isn't needed for optimization. How do you know its one use meets that criterion? It sounds like you're suggesting that `Pgenval` and `Pintval` behave identically outside of calls to `value_kind`? But I don't really think you're saying that, so I'm a bit confused.

Reply:
Right, I'm saying the difference is `layout` gets you a specific `value_kind` instead of just `Pgenval`, which it accomplishes by calling `value_kind`. That is expensive, and sometimes useful, but sometimes not.

As for the particular case, the reason I'm confident this isn't a performance regression is that before the lambda type was changed to use `layout`s instead of `value_kind`s, this was hard coded to be `Pgenval`, as you can see [in the relevant PR](https://github.com/ocaml-flambda/flambda-backend/pull/1032).

A piece of context that probably isn't clear is that that series of PRs added a bunch of calls to `layout`, both in places where the `value_kind` has previously been hardcoded (like this) and in places where `layout`s were newly needed but `value_kind`s had not been (for example, [this PR](https://github.com/ocaml-flambda/flambda-backend/pull/1102) adds layouts to `lapply` and `LSend`).  I suspect these PRs collectively were a noticeable performance regression, but it wasn't measure at the time.  Now that more scrutiny is being applied to the cost of `value_kind`, I think it's useful not to make the problem worse.

Reply:
This has all stemmed from a misunderstanding.

> This should only be used when the precise layout isn't needed for
   optimization

I thought you were talking about a `Layouts.Layout.t`, but you were talking about a `Lambda.layout`. That is, I thought this function was used when we somehow lost the full `Layouts.Layout.t` but just had a sort. So my reaction was "let's just keep the full `Layouts.Layout.t`!" This naturally puzzled you, because you were like "but we don't need a full `Lambda.layout`". Or, in other words, even if we had a `Layouts.Layout.t`, you would just end up calling `sort_of_layout` followed by `layout_of_sort` anyway. You can see why this is confusing!

Bottom line: yes I agree with you. But maybe adding a `Lambda.` in your comment will prevent someone else from falling into the same trap.

Reply:
Drat.  Sorry for the confusion - I've updated the comment.